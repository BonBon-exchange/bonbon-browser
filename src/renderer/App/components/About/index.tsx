import { About } from './About';

export * from './About';
export default About;
