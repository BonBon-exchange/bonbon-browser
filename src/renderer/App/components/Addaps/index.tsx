import { Addaps } from './Addaps';

export * from './Addaps';
export default Addaps;
