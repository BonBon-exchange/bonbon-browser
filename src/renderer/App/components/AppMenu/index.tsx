import { AppMenu } from './AppMenu';

export * from './AppMenu';
export default AppMenu;
