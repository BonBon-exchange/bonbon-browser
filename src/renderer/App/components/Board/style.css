.Board__container {
  display: flex;
  height: 300vh;
  width: calc(100% - 50px);
  top: 0px;
  left: 50px;
  position: absolute;
  background: linear-gradient(
    200.96deg,
    #fedc2a -29.09%,
    #dd5789 51.77%,
    #7a2c9e 129.35%
  );
}
