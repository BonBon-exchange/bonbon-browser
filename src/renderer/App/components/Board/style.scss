#Board__container {
  display: flex;
  min-height: 1000vh !important;
  width: calc(100% - 50px);
  top: 0px;
  left: 50px;
  position: absolute;
  overflow: hidden;
  background: linear-gradient(
    200.96deg,
    #fedc2a -29.09%,
    #dd5789 51.77%,
    #7a2c9e 129.35%
  );

  .Board__edge-snap-left {
    border: 5px dotted var(--color-text-primary);
    left: 60px;
    top: 10px;
    bottom: 10px;
    right: 50%;
    height: calc(100% - 20px);
    position: fixed;
    pointer-events: visible;
    display: none;
    background: var(--color-background-primary);
    z-index: 99;
    opacity: 0.3;
  }

  .Board__edge-snap-right {
    border: 5px dotted var(--color-text-primary);
    left: calc(50% + 50px);
    top: 10px;
    bottom: 10px;
    right: 10px;
    height: calc(100% - 20px);
    position: fixed;
    display: none;
    z-index: 99;
    opacity: 0.3;
    background: var(--color-background-primary);
  }
}

.Board__is-full-size {
  overflow: hidden;
  height: 100% !important;
}
