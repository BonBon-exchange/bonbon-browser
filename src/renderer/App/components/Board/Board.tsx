/* eslint-disable promise/always-return */
/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable import/prefer-default-export */
/* eslint-disable no-use-before-define */
import clsx from 'clsx';
import { AnimatePresence } from 'framer-motion';
import React, { useCallback, useEffect, useState } from 'react';

import { Browser } from 'renderer/App/components/Browser';
import { BrowserProps } from 'renderer/App/components/Browser/Types';
import { useBoard } from 'renderer/App/hooks/useBoard';
import { useBrowserMethods } from 'renderer/App/hooks/useBrowserMethods';
import { useStoreHelpers } from 'renderer/App/hooks/useStoreHelpers';
import { useAppDispatch } from 'renderer/App/store/hooks';
import {
  setActiveBrowser,
  setBoardHeight,
} from 'renderer/App/store/reducers/Board';

import { BoardProps } from './Types';

import './style.scss';

export const Board: React.FC<BoardProps> = ({ isFullSize, boardId }) => {
  const board = useBoard();
  const dispatch = useAppDispatch();
  const { focus } = useBrowserMethods();
  const [items, setItems] = useState<BrowserProps[]>([]);
  const helpers = useStoreHelpers();

  const boardContainer = document.querySelector('#Board__container');

  const makeBrowsers = useCallback((sorted: BrowserProps[]) => {
    return sorted.map((b) => <Browser {...b} key={b.id} firstRendering />);
  }, []);

  const contextMenuListener = (e: MouseEvent) => {
    e.preventDefault();
    window.app.tools.showBoardContextMenu({ x: e.clientX, y: e.clientY });
  };

  useEffect(() => {
    if (!board.isFullSize) {
      setTimeout(() => {
        helpers.board
          .distributeWindowsByOrder(board.browsers)
          .then(() => {
            setTimeout(
              () => board.activeBrowser && focus(board.activeBrowser),
              300
            );
          })
          .catch(console.log);
      }, 300);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [board.isFullSize]);

  useEffect(() => {
    const toSort = [...board.browsers];
    const sorted = toSort.sort((a, b) => {
      return a.id > b.id ? 1 : -1;
    });
    setItems(sorted);

    if (boardContainer) {
      const max = board.browsers.reduce(
        (acc, val) => Math.max(acc, val.top + val.height + 100),
        0
      );

      // @ts-ignore
      boardContainer.style.height = `${Number(max)}px`;

      dispatch(setBoardHeight(max));
    }
  }, [board.browsers, dispatch, boardContainer]);

  useEffect(() => {
    window.app.board.setWindowsCount({
      boardId: board.id,
      count: board.browsers.length,
    });
  }, [board.browsers.length, board.id]);

  useEffect(() => {
    if (board.activeBrowser) {
      focus(board.activeBrowser, true);
      dispatch(setActiveBrowser(board.activeBrowser));
    }
  }, [board.activeBrowser, focus, dispatch]);

  // focus activeBrowser when componentDidMount
  useEffect(() => {
    setTimeout(() => {
      if (board.activeBrowser) focus(board.activeBrowser);
      if (boardId) {
        window.app.browser
          .getUrlToOpen()
          .then((res) => {
            if (res) helpers.browser.add({ url: res });
          })
          .catch(console.log);
      }
    }, 0);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    document
      .getElementById('Board__container')
      ?.addEventListener('contextmenu', contextMenuListener);

    return () =>
      document
        .getElementById('Board__container')
        ?.removeEventListener('contextmenu', contextMenuListener);
  }, []);

  return (
    <div
      id="Board__container"
      className={clsx({
        'Board__is-maximized': board.isFullSize || isFullSize,
        'Board__isnt-maximized': !board.isFullSize && !isFullSize,
      })}
    >
      <AnimatePresence>{makeBrowsers(items)}</AnimatePresence>
      <div className="Board__edge-snap-left" />
      <div className="Board__edge-snap-right" />
      <div className="Board__edge-snap-maximized" />
    </div>
  );
};
