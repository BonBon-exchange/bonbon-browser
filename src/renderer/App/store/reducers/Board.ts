/* eslint-disable @typescript-eslint/no-unused-expressions */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 } from 'uuid';

import { WritableDraft } from 'immer/dist/internal';
import { BoardType } from 'renderer/App/components/Board/Types';
import { BrowserProps } from 'renderer/App/components/Browser/Types';

type UpdateBrowserType = {
  browserId: string;
  params: Partial<BrowserProps>;
};

type UpdateBrowserUrlType = {
  url: string;
  browserId: string;
};

type UpdateBrowserLoadingType = {
  isLoading: boolean;
  browserId: string;
};

type UpdateBrowserTitleType = {
  title: string;
  browserId: string;
};

type UpdateBrowserFavType = {
  favicon: string;
  browserId: string;
};

type UpdateBrowserCertificateFingerprintType = {
  certificateErrorFingerprint: string | null;
  browserId: string;
};

interface BoardState {
  board: BoardType;
}

const boardId = v4();
const newBoard = {
  id: boardId,
  label: 'New collection',
  browsers: [],
  activeBrowser: null,
  closedUrls: [],
  isFullSize: false,
  browsersActivity: [],
  height: 0,
};

export const initialState: BoardState = {
  board: newBoard,
};

const addCLosedUrl = (state: WritableDraft<BoardState>, url: string) => {
  if (state.board.closedUrls.length > 20) {
    state.board.closedUrls.splice(0, 1);
  }
  state.board.closedUrls.push(url);
};

const addBrowserActivity = (
  state: WritableDraft<BoardState>,
  browserId: string
) => {
  if (state.board.browsersActivity.length > 20) {
    state.board.browsersActivity.splice(0, 1);
  }
  state.board.browsersActivity = state.board.browsersActivity.filter(
    (b) => b !== browserId
  );
  state.board.browsersActivity.push(browserId);
};

const findAndSelectTabForExtensionsFromBrowserId = (
  state: WritableDraft<BoardState>,
  browserId: string
): void => {
  const wcId = state.board.browsers.find(
    (b) => b.id === browserId
  )?.webContentsId;
  if (wcId) window.app.browser.select(wcId);
};

const setStateActiveBrowser = (
  state: WritableDraft<BoardState>,
  browserId: string
): void => {
  state.board.activeBrowser = browserId;
  addBrowserActivity(state, browserId);
  findAndSelectTabForExtensionsFromBrowserId(state, browserId);
};

const getBrowserIndexFromBrowserId = (
  state: WritableDraft<BoardState>,
  browserId: string
): number => {
  return state.board.browsers.findIndex((b) => b.id === browserId);
};

export const boardSlice = createSlice({
  name: 'board',
  initialState,
  reducers: {
    setBoard: (state, action: PayloadAction<BoardType>) => {
      state.board = action.payload;
      const activeBrowser = state.board.browsers.find(
        (b) => b.id === state.board.activeBrowser
      );
      if (!state.board.activeBrowser || !activeBrowser) {
        if (state.board.browsers.length > 0) {
          setStateActiveBrowser(
            state,
            state.board.browsers[state.board.browsers.length - 1].id
          );
        }
      }
      if (state.board.activeBrowser) {
        findAndSelectTabForExtensionsFromBrowserId(
          state,
          state.board.activeBrowser
        );
      }
    },
    toggleBoardFullSize: (state) => {
      state.board.isFullSize = !state.board.isFullSize;
      window.app.analytics.event(`toggle_fullsize_${!state.board.isFullSize}`);
    },
    setActiveBrowser: (state, action: PayloadAction<string>) => {
      setStateActiveBrowser(state, action.payload);
    },
    addBrowser: (state, action: PayloadAction<BrowserProps>) => {
      if (action.payload.browserIndex) {
        state.board.browsers.splice(
          action.payload.browserIndex,
          0,
          action.payload
        );
      } else {
        state.board.browsers.push(action.payload);
      }
      setStateActiveBrowser(state, action.payload.id);
      window.app.analytics.event('add_browser');
    },
    setBrowsers: (state, action: PayloadAction<BrowserProps[]>) => {
      state.board.browsers = action.payload;
    },
    updateBrowser: (state, action: PayloadAction<UpdateBrowserType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex] = {
          ...state.board.browsers[browserIndex],
          ...action.payload.params,
        };
      }
    },
    updateBrowserUrl: (state, action: PayloadAction<UpdateBrowserUrlType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].url = action.payload.url;
      }
    },
    updateBrowserLoading: (
      state,
      action: PayloadAction<UpdateBrowserLoadingType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].isLoading = action.payload.isLoading;
      }
    },
    updateBrowserTitle: (
      state,
      action: PayloadAction<UpdateBrowserTitleType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].title = action.payload.title;
      }
    },
    updateBrowserFav: (state, action: PayloadAction<UpdateBrowserFavType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].favicon = action.payload.favicon;
      }
    },
    renameBoard: (state, action: PayloadAction<string>) => {
      state.board.label = action.payload;
      window.app.analytics.event('rename_board');
    },
    minimizeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);
      state.board.browsers[browserIndex].isMinimized = true;
    },
    unminimizeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);

      state.board.browsers[browserIndex].isMinimized = false;
    },
    removeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);

      // remove browser from state
      if (browserIndex > -1) {
        if (!state.board.closedUrls) state.board.closedUrls = [];
        addCLosedUrl(state, state.board.browsers[browserIndex].url);
        state.board.lastClosedBrowserDimensions = [
          state.board.browsers[browserIndex].width,
          state.board.browsers[browserIndex].height,
        ];
        state.board.browsers.splice(browserIndex, 1);

        // clean browsersActivity state
        state.board.browsersActivity = state.board.browsersActivity.filter(
          (b) => b !== action.payload
        );
      }

      // clean activeBrowser
      if (state.board.activeBrowser === action.payload) {
        if (state.board.browsers.length > 0) {
          setStateActiveBrowser(
            state,
            state.board.browsersActivity[
              state.board.browsersActivity.length - 1
            ]
          );
        } else {
          state.board.activeBrowser = null;
          window.app.browser.selectBrowserView();
        }
      }

      window.app.analytics.event('close_browser');
    },
    removeAllBrowsersExcept: (state, action: PayloadAction<string>) => {
      state.board.browsers
        .filter((b) => b.id !== action.payload)
        .forEach((b) => addCLosedUrl(state, b.url));

      state.board.browsers = state.board.browsers.filter(
        (b) => b.id === action.payload
      );

      // clean activeBrowser
      setStateActiveBrowser(state, action.payload);

      // clean browsersActivity state
      state.board.browsersActivity = state.board.browsersActivity.filter(
        (b) => b === action.payload
      );

      // send event
      window.app.analytics.event('close_others_browser');
    },
    updateBrowserCertificateErrorFingerprint: (
      state,
      action: PayloadAction<UpdateBrowserCertificateFingerprintType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].certificateErrorFingerprint =
          action.payload.certificateErrorFingerprint;
      }
    },
    removeAllBrowsers: (state) => {
      if (!state.board.closedUrls) state.board.closedUrls = [];
      state.board.browsers.forEach((b) => addCLosedUrl(state, b.url));
      state.board.browsers = [];
      state.board.browsersActivity = [];
      state.board.activeBrowser = null;
      window.app.browser.selectBrowserView();
      window.app.analytics.event('close_all_browsers');
    },
    removeLastClosedUrl: (state) => {
      state.board.closedUrls.splice(state.board.closedUrls.length - 1, 1);
    },
    toggleSearch: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);
      state.board.browsers[browserIndex].isSearching =
        !state.board.browsers[browserIndex].isSearching;
    },
    setLastResizedBrowserDimensions: (
      state,
      action: PayloadAction<[number, number]>
    ) => {
      state.board.lastResizedBrowserDimensions = action.payload;
    },
    setBoardHeight: (state, action: PayloadAction<number>) => {
      state.board.height = action.payload;
    },
  },
});

export const {
  setBoard,
  updateBrowserUrl,
  updateBrowserFav,
  renameBoard,
  removeBrowser,
  minimizeBrowser,
  unminimizeBrowser,
  removeAllBrowsers,
  addBrowser,
  setActiveBrowser,
  updateBrowserTitle,
  updateBrowser,
  toggleBoardFullSize,
  removeLastClosedUrl,
  setBrowsers,
  removeAllBrowsersExcept,
  updateBrowserLoading,
  updateBrowserCertificateErrorFingerprint,
  toggleSearch,
  setLastResizedBrowserDimensions,
  setBoardHeight,
} = boardSlice.actions;

export default boardSlice.reducer;
