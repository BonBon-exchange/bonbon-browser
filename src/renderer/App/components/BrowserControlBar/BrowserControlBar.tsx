/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable promise/always-return */
/* eslint-disable promise/no-nesting */
/* eslint-disable promise/catch-or-return */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable import/prefer-default-export */
import React, {
  FocusEventHandler,
  KeyboardEventHandler,
  useEffect,
  useState,
  useRef,
} from 'react';
import TextField from '@mui/material/TextField';
import CachedIcon from '@mui/icons-material/Cached';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import ArrowForwardIcon from '@mui/icons-material/ArrowForward';
import HomeIcon from '@mui/icons-material/Home';
import StarBorderIcon from '@mui/icons-material/StarBorder';
import StarIcon from '@mui/icons-material/Star';
import CloseIcon from '@mui/icons-material/Close';

import { BrowserInputSuggestions } from 'renderer/App/components/BrowserInputSuggestions';
import { SuggestionItem } from 'renderer/App/components/BrowserInputSuggestions/Types';
import { updateBrowserUrl } from 'renderer/App/store/reducers/Board';
import { useAppDispatch } from 'renderer/App/store/hooks';
import { isValidHttpUrl, makeSearchUrl } from 'renderer/App/helpers/web';
import { getWebviewFromBrowserId } from 'renderer/App/helpers/dom';
import { useBoard } from 'renderer/App/hooks/useBoard';

import { BrowserControlBarProps } from './Types';

import './style.scss';

export const BrowserControlBar: React.FC<BrowserControlBarProps> = ({
  url,
  browserId,
}) => {
  const [urlInputForSuggestion, setUrlInputForSuggestion] =
    useState<string>(url);
  const [urlInputForAutocomplete, setUrlInputForAutocomplete] =
    useState<string>();
  const [urlInputValue, setUrlInputValue] = useState<string>(url);
  const [showSuggestions, setShowSuggestions] = useState<boolean>(false);
  const [isBookmarked, setIsBookmarked] = useState<boolean>(false);
  const [selectedSuggestion, setSelectedSuggestion] = useState<string | null>(
    null
  );
  const [domainSuggestionResults, setDomainSuggestionResults] = useState<
    SuggestionItem[]
  >([]);
  const [inputCursor, setInputCursor] = useState<number>(0);
  const [userTyped, setUserTyped] = useState<string>('');
  const [userDeleting, setUserDeleting] = useState<boolean>(false);

  const inputRef = useRef<HTMLInputElement>(null);
  const dispatch = useAppDispatch();
  const webview = getWebviewFromBrowserId(browserId);
  const board = useBoard();
  const browser = board.browsers.find((b) => b.id === browserId);

  const hideSuggestions = () => {
    setShowSuggestions(false);
    setSelectedSuggestion(null);
  };

  const handleLoadUserUrl = async (loadUrl: string) => {
    const re =
      /^(http:\/\/www\.|https:\/\/www\.|http:\/\/|https:\/\/)?[a-z0-9]+([-.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/;

    const addHttps =
      !isValidHttpUrl(loadUrl) &&
      isValidHttpUrl(`https://${loadUrl}`) &&
      `https://${loadUrl}`.match(re);

    let newUrl;
    if (addHttps) {
      newUrl = `https://${loadUrl}`;
    } else {
      newUrl = isValidHttpUrl(loadUrl) ? loadUrl : await makeSearchUrl(loadUrl);
    }

    webview?.loadURL(newUrl).catch(console.log);

    dispatch(
      updateBrowserUrl({
        url: newUrl,
        browserId,
      })
    );
  };

  const handleSuggestionClick = (clickedUrl: string) => {
    hideSuggestions();
    handleLoadUserUrl(clickedUrl);
  };

  const urlInputOnKeyUp: KeyboardEventHandler = (e) => {
    if (e.key !== 'ArrowUp') return;
    const target = e.target as HTMLInputElement;
    target?.setSelectionRange(target?.value.length, target?.value.length);
  };

  const urlInputOnKeyDown: KeyboardEventHandler = (e) => {
    const target = e.target as HTMLInputElement;

    if (e.key === 'Escape') {
      setShowSuggestions(false);
      return;
    }

    if (e.key === 'Backspace' || e.key === 'Delete') {
      setUserDeleting(true);
      setUrlInputForAutocomplete(undefined);
      setDomainSuggestionResults([]);
    } else {
      setUserDeleting(false);
    }

    if (selectedSuggestion && showSuggestions && e.key === 'Enter') {
      hideSuggestions();
      handleLoadUserUrl(selectedSuggestion);
      return;
    }

    if (e.code === 'Enter' || e.code === 'NumpadEnter') {
      hideSuggestions();
      domainSuggestionResults[0]
        ? handleLoadUserUrl(
            domainSuggestionResults[0].url ||
              domainSuggestionResults[0].display ||
              ''
          )
        : handleLoadUserUrl(target?.value);
      return;
    }

    setShowSuggestions(true);
  };

  const onFocusInput: FocusEventHandler = (e) => {
    const target = e.target as HTMLInputElement;
    target.select();
  };

  const handleBookmark = () => {
    if (browser?.url && browser?.title) {
      if (isBookmarked) {
        window.app.bookmark.removeBookmark(browser?.url);
        setIsBookmarked(false);
      } else {
        window.app.bookmark
          .addBookmark({
            url: browser?.url,
            name: browser?.title,
          })
          .then(() => {
            setIsBookmarked(true);
          })
          .catch(console.log);
      }
    }
  };

  const goBack = () => {
    webview?.goBack();
  };

  const goForward = () => {
    webview?.goForward();
  };

  const reload = () => {
    webview?.reload();
  };

  const stop = () => {
    webview?.stop();
  };

  const goHome = () => {
    window.app.config.get('browsing.defaultWebpage').then((val) => {
      const defaultWebpage = val as string;
      webview?.loadURL(defaultWebpage).catch(console.log);
      dispatch(
        updateBrowserUrl({
          url: defaultWebpage,
          browserId,
        })
      );
    });
  };

  useEffect(() => {
    if (selectedSuggestion) setUrlInputValue(selectedSuggestion);
  }, [selectedSuggestion]);

  useEffect(() => {
    if (
      domainSuggestionResults[0] &&
      urlInputValue.length > 0 &&
      !userDeleting &&
      selectedSuggestion !== urlInputValue
    )
      setUrlInputForAutocomplete(domainSuggestionResults[0].display);
    else setUrlInputForAutocomplete(undefined);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [domainSuggestionResults]);

  useEffect(() => {
    if (
      urlInputForAutocomplete &&
      urlInputValue.length > 0 &&
      urlInputForAutocomplete.indexOf(urlInputValue) === 0 &&
      selectedSuggestion !== urlInputValue
    )
      setUrlInputValue(urlInputForAutocomplete);

    if (!urlInputForAutocomplete) setUrlInputValue(userTyped);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [urlInputForAutocomplete]);

  useEffect(() => {
    if (
      urlInputValue !== urlInputForAutocomplete &&
      urlInputForAutocomplete &&
      urlInputForAutocomplete.indexOf(urlInputValue) === 0 &&
      urlInputValue !== selectedSuggestion &&
      showSuggestions
    ) {
      setUrlInputValue(urlInputForAutocomplete);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [urlInputValue]);

  useEffect(() => {
    if (urlInputValue !== urlInputForAutocomplete) {
      if (urlInputValue.length < userTyped.length) {
        setUrlInputForAutocomplete(undefined);
      }

      if (urlInputValue !== selectedSuggestion) {
        setInputCursor(urlInputValue.length);
        setUrlInputForSuggestion(urlInputValue);
        setUserTyped(urlInputValue);
      }
    } else {
      const target = inputRef.current?.querySelector('input');
      if (target?.value)
        target?.setSelectionRange(inputCursor, target.value.length);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [urlInputValue]);

  useEffect(() => {
    setUrlInputValue(url);
  }, [url]);

  useEffect(() => {
    window.addEventListener('click', hideSuggestions);
    return () => window.removeEventListener('click', hideSuggestions);
  }, []);

  useEffect(() => {
    if (browser?.url) {
      window.app.bookmark
        .isBookmarked(browser?.url)
        .then((val) => setIsBookmarked(val))
        .catch(console.log);
    }
  }, [browser?.url]);

  useEffect(() => {
    if (urlInputValue.length === 0) hideSuggestions();
  }, [urlInputValue.length]);

  return (
    <div className="BrowserControlBar__container">
      <div className="BrowserControlBar__controls">
        <div className="BrowserControlBar__control" onClick={goBack}>
          <ArrowBackIcon />
        </div>
        <div className="BrowserControlBar__control" onClick={goForward}>
          <ArrowForwardIcon />
        </div>
        {browser?.isLoading ? (
          <div className="BrowserControlBar__control" onClick={stop}>
            <CloseIcon />
          </div>
        ) : (
          <div className="BrowserControlBar__control" onClick={reload}>
            <CachedIcon />
          </div>
        )}
        <div className="BrowserControlBar__control" onClick={goHome}>
          <HomeIcon />
        </div>
      </div>
      <TextField
        variant="standard"
        value={urlInputValue}
        className="BrowserControlBar_url-input"
        onKeyDown={urlInputOnKeyDown}
        onKeyUp={urlInputOnKeyUp}
        onFocus={onFocusInput}
        ref={inputRef}
        onChange={(e) => setUrlInputValue(e.target.value)}
      />
      <div
        className="BrowserControlBar__bookmark-control"
        onClick={handleBookmark}
      >
        {isBookmarked ? <StarIcon /> : <StarBorderIcon />}
      </div>
      {showSuggestions && (
        <BrowserInputSuggestions
          inputValue={urlInputForSuggestion}
          handleSuggestionClick={handleSuggestionClick}
          setSelectedSuggestion={setSelectedSuggestion}
          setDomainSuggestionResults={setDomainSuggestionResults}
        />
      )}
    </div>
  );
};
