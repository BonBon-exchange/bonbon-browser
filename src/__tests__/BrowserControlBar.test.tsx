/* eslint-disable react/jsx-props-no-spreading */
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { Middleware } from '@reduxjs/toolkit';

import { BrowserControlBar } from '../renderer/App/components/BrowserControlBar';
import { mockWindow } from './beforeAll';
import { initialState } from '../renderer/App/store/reducers/Board';

let store: any;
const middlewares: Middleware[] = [];

const browserControlBarProps = {
  goBack: jest.fn(),
  goForward: jest.fn(),
  reload: jest.fn(),
  url: 'https://www.github.com',
  browserId: '123abc',
  goHome: jest.fn(),
};

describe('BrowserControlBar', () => {
  beforeAll(() => {
    mockWindow();
    const mockStore = configureStore(middlewares);
    store = mockStore({ board: initialState });
  });

  it('should render', () => {
    expect(
      render(
        <Provider store={store}>
          <BrowserControlBar {...browserControlBarProps} />
        </Provider>
      )
    ).toBeTruthy();
  });
});
