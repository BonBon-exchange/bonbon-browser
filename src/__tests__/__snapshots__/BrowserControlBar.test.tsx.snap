// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserControlBar should render 1`] = `
<div
  className="BrowserControlBar__container"
>
  <div
    className="BrowserControlBar__controls"
  >
    <div
      className="BrowserControlBar__control"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
    </div>
    <div
      className="BrowserControlBar__control"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowForwardIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
        />
      </svg>
    </div>
    <div
      className="BrowserControlBar__control"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CachedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m19 8-4 4h3c0 3.31-2.69 6-6 6-1.01 0-1.97-.25-2.8-.7l-1.46 1.46C8.97 19.54 10.43 20 12 20c4.42 0 8-3.58 8-8h3l-4-4zM6 12c0-3.31 2.69-6 6-6 1.01 0 1.97.25 2.8.7l1.46-1.46C15.03 4.46 13.57 4 12 4c-4.42 0-8 3.58-8 8H1l4 4 4-4H6z"
        />
      </svg>
    </div>
    <div
      className="BrowserControlBar__control"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="HomeIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root BrowserControlBar_url-input css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <div
      className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        disabled={false}
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="https://www.github.com"
      />
    </div>
  </div>
  <div
    className="BrowserControlBar__bookmark-control"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="StarBorderIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m22 9.24-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
      />
    </svg>
  </div>
</div>
`;
